src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegion.java

```diff
@@ -0,0 +1,609 @@
+     * @param activeProbThreshold minimum probability for a site to be considered active, as reported by the provided evaluator
+     * @param maxProbPropagationDistance maximum number of bases probabilities can propagate in each direction when finding region boundaries
+     * @return a Iterable over one or more AssemblyRegions, each marked as either "active" or "inactive", spanning
+     *         part of the provided ReadShard, and filled with all reads that overlap the region.
+     */
+    public static Iterable<AssemblyRegion> createFromReadShard( final ReadShard shard,
+                                                                final SAMFileHeader readsHeader,
+                                                                final ReferenceContext referenceContext,
+                                                                final FeatureContext features,
+                                                                final AssemblyRegionEvaluator evaluator,
+                                                                final int minRegionSize,
+                                                                final int maxRegionSize,
+                                                                final int assemblyRegionPadding,
+                                                                final double activeProbThreshold,
+                                                                final int maxProbPropagationDistance ) {
+        Utils.nonNull(shard);
+        Utils.nonNull(readsHeader);
+        Utils.nonNull(referenceContext);
+        Utils.nonNull(features);
+        Utils.nonNull(evaluator);
+        Utils.validateArg(minRegionSize >= 1, "minRegionSize must be >= 1");
+        Utils.validateArg(maxRegionSize >= 1, "maxRegionSize must be >= 1");
+        Utils.validateArg(minRegionSize <= maxRegionSize, "minRegionSize must be <= maxRegionSize");
+        Utils.validateArg(assemblyRegionPadding >= 0, "assemblyRegionPadding must be >= 0");
+        Utils.validateArg(activeProbThreshold >= 0.0, "activeProbThreshold must be >= 0.0");
+        Utils.validateArg(maxProbPropagationDistance >= 0, "maxProbPropagationDistance must be >= 0");
+
+        // TODO: refactor this method so that we don't need to load all reads from the shard into memory at once!
+        final List<GATKRead> windowReads = new ArrayList<>();
+        for ( final GATKRead read : shard ) {
+            windowReads.add(read);
+        }
+
+        final LocusIteratorByState locusIterator = new LocusIteratorByState(windowReads.iterator(), DownsamplingMethod.NONE, false, false, ReadUtils.getSamplesFromHeader(readsHeader), readsHeader);
+        final ActivityProfile activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readsHeader);
+
+        // First, use our activity profile to determine the bounds of each assembly region:
```

Comment:
extract a new method from this part of the createFromReadShard method (how about `determineRegionBounds`?)